Fix default no token mode for older JAX versions

macekmar opened the pull request:
If JAX < 0.5.1, the default no token mode does not work. See issue #274 and:
  - https://github.com/mpi4jax/mpi4jax/issues/274
  - https://github.com/jax-ml/jax/issues/26087

This commit sets that if JAX < 0.5.1, it switches to token mode. However, if the user sets "MPI4JAX_PREFER_NOTOKEN=1" it keeps it becuase, AFAIK, the problem is only limited to JAX linear algebra.

dionhaefner commented on mpi4jax/_src/utils.py:
```diff
@@ -176,12 +178,31 @@ def has_sycl_support() -> bool:
+    # No-token mode only works with JAX >= 0.5.1. See:
+    # - https://github.com/mpi4jax/mpi4jax/issues/274
+    # - https://github.com/jax-ml/jax/issues/26087
+    jax_ver_sufficient = versiontuple(jax.__version__) >= (0, 5, 1)
```
We need to start earlier:

```suggestion
    jax_ver_sufficient = versiontuple(jax.__version__) >= (0, 5, 0)
```

`v0.5.0` doesn't work with tokens (see #273).

dionhaefner replied:
(Let's not forget to update the references to 0.5.1 elsewhere too)

macekmar replied:
No, it has to be `jax_ver_sufficient >= (0,5,1)` since 0.5.0 still has a bug.

> (Let's not forget to update the references to 0.5.1 elsewhere too)
I have found two places where we reference JAX version:
1. setup.py
We have `JAX_MINIMUM_VERSION = "0.4.35"` but if we set it to 0.5.1 it will not install it on Intel Macs, right?
One could issue this patch and once you drop the token mode completely also increase the minimum JAX version.

2. jax_compat.py
Function `check_jax_version()` only issues a warning if JAX version is too high.
Perhaps we could add a warning for users with JAX versions below 0.5.1, informing them that we will default to token mode and that in the future support for versions below 0.5.1 will be discontinued?
However, if the user is already on the latest available but outdated version, the warning is not helpful.

dionhaefner replied:
I get that v0.5.0 is buggy when using with custom linear solves. However, **all versions past v0.5.0 do not work at all with notoken disabled**.  So `jax_ver_sufficient` must use `(0, 5, 0)`, and all other references to v0.5.1 _within this function_ should be updated.

PhilipVinc replied:
yeah, I agree with @dionhaefner . 
I think the minimum jax version should be 0.5.0. 
(@macekmar netket+mpi is broken in jax 0.5.0, there's nothing you can do. It's a netket problem, not a mpi4jax problem)

PhilipVinc replied:
If we have a test using custom linear solvers, we can simply skip it if jax==0.5.0

dionhaefner commented on mpi4jax/_src/utils.py:
```diff
@@ -176,12 +178,31 @@ def has_sycl_support() -> bool:
+        use_notoken = jax_ver_sufficient
+    else:
+        use_notoken = env_var.lower() in ("1", "true", "on")
+
+    if use_notoken is False and env_var is None:
+        warnings.warn(
+            "Defaulting to the deprecated token mode. "
+            "No-token mode fully works only for JAX version >= 0.5.1 but you have {jax.__version__}."
+            "You can set the environment variable `MPI4JAX_PREFER_NOTOKEN=0` to silence this warning."
+        )
+
+    if use_notoken and not jax_ver_sufficient:
+        warnings.warn(
+            f"No-token mode fully works only for JAX version >= 0.5.1 but you have {jax.__version__}."
+        )
```
I reckon those will be more confusing than helpful. If people are using an older JAX, silently using token mode is fine since that's the only option we have. Similarly, people using an older JAX + notoken mode have already explicitly opted into this functionality so should be assumed they know what they're doing (and 99% of functionality does indeed work).

```suggestion
```

macekmar replied:
Agreed.
I also have to update the tests, but first see the 2nd point (jax_compat) above.

macekmar commented:
I did the changes as suggested above (0.5.1→0.5.0, removed warnings, added Vicentini's MWE which is skipped if JAX=0.5.0)
On my side:

- tests fail if JAX>=0.5.0 and MPI4JAX_PREFER_NOTOKEN=0
- JAX <= 0.4.38 and MPI4JAX_PREFER_NOTOKEN=1 (Vicentini's MWE).

I have no idea why test_barrier.py is not passing above.

dionhaefner commented on tests/test_prefer_notoken.py:
```diff
@@ -0,0 +1,80 @@
+    if jax.__version__ == "0.5.0":
+        # We need no-token mode for JAX>=0.5.0 but the code below
+        # does not work for JAX<0.5.1 in no-token mode.
+        return
```
Let's use [xfail](https://docs.pytest.org/en/stable/how-to/skipping.html#xfail-mark-test-functions-as-expected-to-fail) here instead.

dionhaefner reviewed (approved):
Thanks @macekmar

dionhaefner commented:
If it passes then by accident, since the test you pushed doesn't use tokens correctly to enforce order.
